In [1]:
# Installation
! pip install smolagents
# To install from source instead of the last release, comment the command above and uncomment the following one.
# ! pip install git+https://github.com/huggingface/smolagents.git

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.8/149.8 kB 3.5 MB/s eta 0:00:00


# Text-to-SQL

In this tutorial, we’ll see how to implement an agent that leverages SQL using `smolagents`.

> Let's start with the golden question: why not keep it simple and use a standard text-to-SQL pipeline?

A standard text-to-sql pipeline is brittle, since the generated SQL query can be incorrect. Even worse, the query could be incorrect, but not raise an error, instead giving some incorrect/useless outputs without raising an alarm.

👉 Instead, an agent system is able to critically inspect outputs and decide if the query needs to be changed or not, thus giving it a huge performance boost.

Let’s build this agent! 💪

Run the line below to install required dependencies:
```bash
!pip install smolagents python-dotenv sqlalchemy --upgrade -q
```
To call Inference Providers, you will need a valid token as your environment variable `HF_TOKEN`.
We use python-dotenv to load it.

In [2]:
from dotenv import load_dotenv
load_dotenv()

False

Then, we setup the SQL environment:

In [3]:
from sqlalchemy import (
    create_engine,
    MetaData,
    Table,
    Column,
    String,
    Integer,
    Float,
    insert,
    inspect,
    text,
)

engine = create_engine("sqlite:///:memory:")
metadata_obj = MetaData()

def insert_rows_into_table(rows, table, engine=engine):
    for row in rows:
        stmt = insert(table).values(**row)
        with engine.begin() as connection:
            connection.execute(stmt)

table_name = "receipts"
receipts = Table(
    table_name,
    metadata_obj,
    Column("receipt_id", Integer, primary_key=True),
    Column("customer_name", String(16), primary_key=True),
    Column("price", Float),
    Column("tip", Float),
)
metadata_obj.create_all(engine)

rows = [
    {"receipt_id": 1, "customer_name": "Alan Payne", "price": 12.06, "tip": 1.20},
    {"receipt_id": 2, "customer_name": "Alex Mason", "price": 23.86, "tip": 0.24},
    {"receipt_id": 3, "customer_name": "Woodrow Wilson", "price": 53.43, "tip": 5.43},
    {"receipt_id": 4, "customer_name": "Margaret James", "price": 21.11, "tip": 1.00},
]
insert_rows_into_table(rows, receipts)

### Build our agent

Now let’s make our SQL table retrievable by a tool.

The tool’s description attribute will be embedded in the LLM’s prompt by the agent system: it gives the LLM information about how to use the tool. This is where we want to describe the SQL table.

In [4]:
inspector = inspect(engine)
columns_info = [(col["name"], col["type"]) for col in inspector.get_columns("receipts")]

table_description = "Columns:\n" + "\n".join([f"  - {name}: {col_type}" for name, col_type in columns_info])
print(table_description)

Columns:
  - receipt_id: INTEGER
  - customer_name: VARCHAR(16)
  - price: FLOAT
  - tip: FLOAT


```text
Columns:
  - receipt_id: INTEGER
  - customer_name: VARCHAR(16)
  - price: FLOAT
  - tip: FLOAT
```

Now let’s build our tool. It needs the following: (read [the tool doc](https://huggingface.co/docs/smolagents/main/en/examples/../tutorials/tools) for more detail)
- A docstring with an `Args:` part listing arguments.
- Type hints on both inputs and output.

In [5]:
from smolagents import tool

@tool
def sql_engine(query: str) -> str:
    """
    Allows you to perform SQL queries on the table. Returns a string representation of the result.
    The table is named 'receipts'. Its description is as follows:
        Columns:
        - receipt_id: INTEGER
        - customer_name: VARCHAR(16)
        - price: FLOAT
        - tip: FLOAT

    Args:
        query: The query to perform. This should be correct SQL.
    """
    output = ""
    with engine.connect() as con:
        rows = con.execute(text(query))
        for row in rows:
            output += "\n" + str(row)
    return output

Now let us create an agent that leverages this tool.

We use the `CodeAgent`, which is smolagents’ main agent class: an agent that writes actions in code and can iterate on previous output according to the ReAct framework.

The model is the LLM that powers the agent system. `InferenceClientModel` allows you to call LLMs using HF’s Inference API, either via Serverless or Dedicated endpoint, but you could also use any proprietary API.

In [6]:
from smolagents import CodeAgent, InferenceClientModel

agent = CodeAgent(
    tools=[sql_engine],
    model=InferenceClientModel(model_id="meta-llama/Llama-3.1-8B-Instruct"),
)
agent.run("Can you give me the name of the client who got the most expensive receipt?")

╭──────────────────────────────────────────────────── New run ────────────────────────────────────────────────────╮
│                                                                                                                 │
│ Can you give me the name of the client who got the most expensive receipt?                                      │
│                                                                                                                 │
╰─ InferenceClientModel - meta-llama/Llama-3.1-8B-Instruct ───────────────────────────────────────────────────────╯

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  max_price_receipt = sql_engine("SELECT customer_name, MAX(price + tip) as max_price FROM receipts")              
  max_price = float(max_price_receipt.split("MAX(price + tip) as max_price ")[1].split(" FROM")[0])                
  print(f"The maximum price is {max_price}.")                                                                      
  max_customer_name = sql_engine(f"SELECT customer_name FROM receipts WHERE price + tip = {max_price}")            
  print(f"The customer with the maximum price is {max_customer_name}")                                             
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'max_price = float(max_price_receipt.split("MAX(price + tip) as max_price 
")[1\].split(" FROM")[0\])' due to: InterpreterError: Could not index ["\n('Woodrow Wilson', 58.86)"\] with '1': 
IndexError: list index out of range

[Step 1: Duration 4.11 seconds| Input tokens: 2,117 | Output tokens: 142]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 2 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  max_price_receipt = sql_engine("SELECT customer_name, price + tip FROM receipts ORDER BY price + tip DESC LIMIT  
  1")                                                                                                              
  max_customer_name, max_price = max_price_receipt.split(" (")[:-1].split(", ")                                    
  print(f"The customer with the maximum price is {max_customer_name} and the maximum price is {max_price}.")       
  final_answer(max_customer_name)                                                                                  
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'max_customer_name, max_price = max_price_receipt.split(" (")[:-1\].split(", ")' due 
to: InterpreterError: Object [] has no attribute split

[Step 2: Duration 3.66 seconds| Input tokens: 4,628 | Output tokens: 285]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 3 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  max_price = sql_engine("SELECT MAX(price + tip) FROM receipts")                                                  
  max_customer_name = sql_engine(f"SELECT customer_name FROM receipts WHERE price + tip = {max_price}")            
  final_answer(max_customer_name)                                                                                  
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'max_customer_name = sql_engine(f"SELECT customer_name FROM receipts WHERE price + 
tip = {max_price}")' due to: OperationalError: (sqlite3.OperationalError) near ")": syntax error
[SQL: SELECT customer_name FROM receipts WHERE price + tip = 
(58.86,)\]
(Background on this error at: https://sqlalche.me/e/20/e3q8)

[Step 3: Duration 3.24 seconds| Input tokens: 7,485 | Output tokens: 401]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 4 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  max_price = sql_engine("SELECT MAX(price + tip) FROM receipts")                                                  
  max_price_str = str(max_price).strip("(").strip(")").strip(" ")                                                  
  max_customer_name = sql_engine(f"SELECT customer_name FROM receipts WHERE price + tip = {max_price_str}")        
  final_answer(max_customer_name)                                                                                  
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'max_customer_name = sql_engine(f"SELECT customer_name FROM receipts WHERE price + 
tip = {max_price_str}")' due to: OperationalError: (sqlite3.OperationalError) incomplete input
[SQL: SELECT customer_name FROM receipts WHERE price + tip = 
(58.86,\]
(Background on this error at: https://sqlalche.me/e/20/e3q8)

[Step 4: Duration 3.97 seconds| Input tokens: 10,676 | Output tokens: 570]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 5 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  result = sql_engine("SELECT customer_name FROM receipts ORDER BY price + tip DESC LIMIT 1")                      
  final_answer(result)                                                                                             
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Final answer: 
('Woodrow Wilson',)

[Step 5: Duration 2.53 seconds| Input tokens: 14,272 | Output tokens: 662]

"\n('Woodrow Wilson',)"

### Level 2: Table joins

Now let’s make it more challenging! We want our agent to handle joins across multiple tables.

So let’s make a second table recording the names of waiters for each receipt_id!

In [7]:
table_name = "waiters"
waiters = Table(
    table_name,
    metadata_obj,
    Column("receipt_id", Integer, primary_key=True),
    Column("waiter_name", String(16), primary_key=True),
)
metadata_obj.create_all(engine)

rows = [
    {"receipt_id": 1, "waiter_name": "Corey Johnson"},
    {"receipt_id": 2, "waiter_name": "Michael Watts"},
    {"receipt_id": 3, "waiter_name": "Michael Watts"},
    {"receipt_id": 4, "waiter_name": "Margaret James"},
]
insert_rows_into_table(rows, waiters)

Since we changed the table, we update the `SQLExecutorTool` with this table’s description to let the LLM properly leverage information from this table.

In [8]:
updated_description = """Allows you to perform SQL queries on the table. Beware that this tool's output is a string representation of the execution output.
It can use the following tables:"""

inspector = inspect(engine)
for table in ["receipts", "waiters"]:
    columns_info = [(col["name"], col["type"]) for col in inspector.get_columns(table)]

    table_description = f"Table '{table}':\n"

    table_description += "Columns:\n" + "\n".join([f"  - {name}: {col_type}" for name, col_type in columns_info])
    updated_description += "\n\n" + table_description

print(updated_description)

Allows you to perform SQL queries on the table. Beware that this tool's output is a string representation of the execution output.
It can use the following tables:

Table 'receipts':
Columns:
  - receipt_id: INTEGER
  - customer_name: VARCHAR(16)
  - price: FLOAT
  - tip: FLOAT

Table 'waiters':
Columns:
  - receipt_id: INTEGER
  - waiter_name: VARCHAR(16)


Since this request is a bit harder than the previous one, we’ll switch the LLM engine to use the more powerful [Qwen/Qwen3-Next-80B-A3B-Thinking](https://huggingface.co/Qwen/Qwen3-Next-80B-A3B-Thinking)!

In [18]:
sql_engine.description = updated_description

agent = CodeAgent(
    tools=[sql_engine],
    model=InferenceClientModel(model_id="Qwen/Qwen3-Next-80B-A3B-Thinking"),
)

agent.run("Which waiter got more total money from tips?")

╭──────────────────────────────────────────────────── New run ────────────────────────────────────────────────────╮
│                                                                                                                 │
│ Which waiter got more total money from tips?                                                                    │
│                                                                                                                 │
╰─ InferenceClientModel - Qwen/Qwen3-Next-80B-A3B-Thinking ───────────────────────────────────────────────────────╯

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

[Step 1: Duration 9.55 seconds]

KeyboardInterrupt: 

It directly works! The setup was surprisingly simple, wasn’t it?

This example is done! We've touched upon these concepts:
- Building new tools.
- Updating a tool's description.
- Switching to a stronger LLM helps agent reasoning.

✅ Now you can go build this text-to-SQL system you’ve always dreamt of! ✨

# Task
Use the dataset in the file "/content/sample.zip" to answer questions.

## Inspect the provided data

### Subtask:
Extract the contents of the zip file and identify the data file(s) and their format (e.g., CSV, JSON, etc.).


**Reasoning**:
Extract the contents of the zip file and identify the data file(s) and their format.



In [32]:
import zipfile
import os

zip_file_path = '/content/sample.zip'
extract_dir = '/content/extracted_data'
os.makedirs(extract_dir, exist_ok=True)

with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_dir)

extracted_files = os.listdir(extract_dir)
print("Files extracted from the zip file:")
for file_name in extracted_files:
    print(file_name)

data_files = [f for f in extracted_files if '.' in f]
print("\nIdentified data files and their inferred formats:")
for data_file in data_files:
    name, ext = os.path.splitext(data_file)
    print(f"- {data_file}: {ext.lstrip('.').upper()}")

Files extracted from the zip file:
dbo.Orders.csv
dbo.Customers.columns
dbo.Products.columns
dbo.Products.csv
dbo.Customers.csv
dbo.Shippers.columns
dbo.Suppliers.csv
dbo.Order Details.csv
dbo.Order Details.columns
dbo.Employees.columns
dbo.Suppliers.columns
dbo.Shippers.csv
dbo.Employees.csv
dbo.Categories.columns
dbo.Categories.csv
dbo.Orders.columns

Identified data files and their inferred formats:
- dbo.Orders.csv: CSV
- dbo.Customers.columns: COLUMNS
- dbo.Products.columns: COLUMNS
- dbo.Products.csv: CSV
- dbo.Customers.csv: CSV
- dbo.Shippers.columns: COLUMNS
- dbo.Suppliers.csv: CSV
- dbo.Order Details.csv: CSV
- dbo.Order Details.columns: COLUMNS
- dbo.Employees.columns: COLUMNS
- dbo.Suppliers.columns: COLUMNS
- dbo.Shippers.csv: CSV
- dbo.Employees.csv: CSV
- dbo.Categories.columns: COLUMNS
- dbo.Categories.csv: CSV
- dbo.Orders.columns: COLUMNS


## Load the data

### Subtask:
Load the data from the identified file(s) into a suitable format (e.g., pandas DataFrame).


**Reasoning**:
Import the pandas library and load the CSV files into a dictionary of dataframes.



In [31]:
import pandas as pd

extracted_dir = '/content/extracted_data'
extracted_files = os.listdir(extracted_dir)

csv_files = [f for f in extracted_files if f.endswith('.csv')]

dataframes = {}
for csv_file in csv_files:
    file_path = os.path.join(extracted_dir, csv_file)
    df_name = os.path.splitext(csv_file)[0]
    dataframes[df_name] = pd.read_csv(file_path)

print("Loaded tables (DataFrames):")
print(dataframes.keys())

Loaded tables (DataFrames):
dict_keys(['dbo.Orders', 'dbo.Products', 'dbo.Customers', 'dbo.Suppliers', 'dbo.Order Details', 'dbo.Shippers', 'dbo.Employees', 'dbo.Categories'])


## Adapt the database setup

### Subtask:
Modify the existing database setup code to create tables and insert data from your loaded dataset. This might involve inspecting the columns and data types of your data.


**Reasoning**:
Drop existing tables, create new tables based on the loaded dataframes, and insert data into them.



In [30]:
metadata_obj.drop_all(engine)
metadata_obj = MetaData()

for table_name, df in dataframes.items():
    columns = [Column(col, dtype) for col, dtype in zip(df.columns, [String if dt == 'object' else Float if dt == 'float64' else Integer for dt in df.dtypes])]
    table = Table(table_name, metadata_obj, *columns)
    metadata_obj.create_all(engine)
    df.to_sql(table_name, engine, if_exists='append', index=False)

inspector = inspect(engine)
print("Tables created in the database:")
print(inspector.get_table_names())

Tables created in the database:
['dbo.Categories', 'dbo.Customers', 'dbo.Employees', 'dbo.Order Details', 'dbo.Orders', 'dbo.Products', 'dbo.Shippers', 'dbo.Suppliers']


## Update the tool description

### Subtask:
Update the `sql_engine` tool's description to accurately reflect the tables and columns in your dataset.


**Reasoning**:
Initialize the updated_description string and get the inspector object.



In [29]:
updated_description = """Allows you to perform SQL queries on the table. Beware that this tool's output is a string representation of the execution output.
It can use the following tables:"""

inspector = inspect(engine)

**Reasoning**:
Iterate through the tables, get column information, format the description, and append it to the updated_description string.



In [28]:
for table in inspector.get_table_names():
    columns_info = [(col["name"], col["type"]) for col in inspector.get_columns(table)]

    table_description = f"\n\nTable '{table}':\n"
    table_description += "Columns:\n" + "\n".join([f"  - {name}: {col_type}" for name, col_type in columns_info])

    updated_description += table_description

**Reasoning**:
Update the tool's description and print the updated description to verify.



In [27]:
sql_engine.description = updated_description
print(updated_description)

Allows you to perform SQL queries on the table. Beware that this tool's output is a string representation of the execution output.
It can use the following tables:

Table 'dbo.Categories':
Columns:
  - CategoryID: INTEGER
  - CategoryName: VARCHAR
  - Description: VARCHAR

Table 'dbo.Customers':
Columns:
  - CustomerID: VARCHAR
  - CompanyName: VARCHAR
  - ContactName: VARCHAR
  - ContactTitle: VARCHAR
  - Address: VARCHAR
  - City: VARCHAR
  - Region: VARCHAR
  - PostalCode: VARCHAR
  - Country: VARCHAR
  - Phone: VARCHAR
  - Fax: VARCHAR

Table 'dbo.Employees':
Columns:
  - EmployeeID: INTEGER
  - LastName: VARCHAR
  - FirstName: VARCHAR
  - Title: VARCHAR
  - TitleOfCourtesy: VARCHAR
  - BirthDate: VARCHAR
  - HireDate: VARCHAR
  - Address: VARCHAR
  - City: VARCHAR
  - Region: VARCHAR
  - PostalCode: VARCHAR
  - Country: VARCHAR
  - HomePhone: VARCHAR
  - Extension: INTEGER
  - Photo: VARCHAR
  - Notes: VARCHAR
  - ReportsTo: FLOAT
  - PhotoPath: VARCHAR

Table 'dbo.Order Details':

## Adjust the agent's query

### Subtask:
Modify the agent's query to match the structure and content of your data, based on the question you want to ask.


**Reasoning**:
Define a new SQL query to answer a relevant question based on the available tables and then update the agent to run this query.



In [16]:
new_query = "SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM dbo.Customers c JOIN dbo.\"Orders\" o ON c.CustomerID = o.CustomerID JOIN dbo.\"Order Details\" od ON o.OrderID = od.OrderID GROUP BY c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;"

agent = CodeAgent(
    tools=[sql_engine],
    model=InferenceClientModel(model_id="Qwen/Qwen3-Next-80B-A3B-Thinking"),
)

agent.run(f"Find the top 10 customers by total revenue using the following query: {new_query}")

╭──────────────────────────────────────────────────── New run ────────────────────────────────────────────────────╮
│                                                                                                                 │
│ Find the top 10 customers by total revenue using the following query: SELECT c.CompanyName, SUM(od.UnitPrice *  │
│ od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM dbo.Customers c JOIN dbo."Orders" o ON c.CustomerID =     │
│ o.CustomerID JOIN dbo."Order Details" od ON o.OrderID = od.OrderID GROUP BY c.CompanyName ORDER BY TotalRevenue │
│ DESC LIMIT 10;                                                                                                  │
│                                                                                                                 │
╰─ InferenceClientModel - Qwen/Qwen3-Next-80B-A3B-Thinking ───────────────────────────────────────────────────────╯

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  query = 'SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM          
  dbo.Customers c JOIN dbo."Orders" o ON c.CustomerID = o.CustomerID JOIN dbo."Order Details" od ON o.OrderID =    
  od.OrderID GROUP BY c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;'                                          
  result = sql_engine(query=query)                                                                                 
  print(result)                                                                                                    
  final_answer(result)                                                                                             
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'result = sql_engine(query=query)' due to: OperationalError: 
(sqlite3.OperationalError) no such table: dbo.Customers
[SQL: SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM dbo.Customers 
c JOIN dbo."Orders" o ON c.CustomerID = o.CustomerID JOIN dbo."Order Details" od ON o.OrderID = od.OrderID GROUP BY
c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;\]
(Background on this error at: https://sqlalche.me/e/20/e3q8)

[Step 1: Duration 12.00 seconds| Input tokens: 2,766 | Output tokens: 2,149]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 2 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  query = 'SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM          
  Customers c JOIN "Orders" o ON c.CustomerID = o.CustomerID JOIN "Order Details" od ON o.OrderID = od.OrderID     
  GROUP BY c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;'                                                     
  result = sql_engine(query=query)                                                                                 
  print(result)                                                                                                    
  final_answer(result)                                                                                             
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'result = sql_engine(query=query)' due to: OperationalError: 
(sqlite3.OperationalError) no such table: Customers
[SQL: SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM Customers c 
JOIN "Orders" o ON c.CustomerID = o.CustomerID JOIN "Order Details" od ON o.OrderID = od.OrderID GROUP BY 
c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;\]
(Background on this error at: https://sqlalche.me/e/20/e3q8)

[Step 2: Duration 7.21 seconds| Input tokens: 5,982 | Output tokens: 3,353]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 3 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  query = 'SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM          
  customers c JOIN orders o ON c.CustomerID = o.CustomerID JOIN order_details od ON o.OrderID = od.OrderID GROUP   
  BY c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;'                                                           
  result = sql_engine(query=query)                                                                                 
  print(result)                                                                                                    
  final_answer(result)                                                                                             
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Code execution failed at line 'result = sql_engine(query=query)' due to: OperationalError: 
(sqlite3.OperationalError) no such table: customers
[SQL: SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM customers c 
JOIN orders o ON c.CustomerID = o.CustomerID JOIN order_details od ON o.OrderID = od.OrderID GROUP BY c.CompanyName
ORDER BY TotalRevenue DESC LIMIT 10;\]
(Background on this error at: https://sqlalche.me/e/20/e3q8)

[Step 3: Duration 29.49 seconds| Input tokens: 9,642 | Output tokens: 8,756]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 4 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  query = "SELECT name FROM sqlite_master WHERE type='table';"                                                     
  result = sql_engine(query=query)                                                                                 
  print(result)                                                                                                    
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Execution logs:

('dbo.Orders',)
('dbo.Products',)
('dbo.Customers',)
('dbo.Suppliers',)
('dbo.Order Details',)
('dbo.Shippers',)
('dbo.Employees',)
('dbo.Categories',)

Out: None

[Step 4: Duration 34.84 seconds| Input tokens: 13,748 | Output tokens: 15,041]

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 5 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

─ Executing parsed code: ──────────────────────────────────────────────────────────────────────────────────────── 
  query = 'SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM          
  "dbo.Customers" c JOIN "dbo.Orders" o ON c.CustomerID = o.CustomerID JOIN "dbo.Order Details" od ON o.OrderID =  
  od.OrderID GROUP BY c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;'                                          
  result = sql_engine(query=query)                                                                                 
  print(result)                                                                                                    
  final_answer(result)                                                                                             
 ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Execution logs:

('Suprêmes délices', 3597.9)
('Frankenversand', 3536.6)
('Ernst Handel', 3488.6800000000003)
('Hanari Carnes', 2997.4)
('QUICK-Stop', 2555.68)
('Richter Supermarkt', 2490.5)
('Berglunds snabbköp', 2102.0)
('Rattlesnake Canyon Grocery', 2040.0)
('Toms Spezialitäten', 1863.4)
('Wartian Herkku', 1722.56)

Final answer: 
('Suprêmes délices', 3597.9)
('Frankenversand', 3536.6)
('Ernst Handel', 3488.6800000000003)
('Hanari Carnes', 2997.4)
('QUICK-Stop', 2555.68)
('Richter Supermarkt', 2490.5)
('Berglunds snabbköp', 2102.0)
('Rattlesnake Canyon Grocery', 2040.0)
('Toms Spezialitäten', 1863.4)
('Wartian Herkku', 1722.56)

[Step 5: Duration 15.96 seconds| Input tokens: 18,058 | Output tokens: 17,695]

"\n('Suprêmes délices', 3597.9)\n('Frankenversand', 3536.6)\n('Ernst Handel', 3488.6800000000003)\n('Hanari Carnes', 2997.4)\n('QUICK-Stop', 2555.68)\n('Richter Supermarkt', 2490.5)\n('Berglunds snabbköp', 2102.0)\n('Rattlesnake Canyon Grocery', 2040.0)\n('Toms Spezialitäten', 1863.4)\n('Wartian Herkku', 1722.56)"

Here are 20 SQL queries to test the agent:

1.  Find the total number of orders.
    `SELECT COUNT(*) FROM "dbo.Orders";`
2.  List the names of all products.
    `SELECT ProductName FROM "dbo.Products";`
3.  Find the company names of all customers in 'London'.
    `SELECT CompanyName FROM "dbo.Customers" WHERE City = 'London';`
4.  Get the distinct countries of all suppliers.
    `SELECT DISTINCT Country FROM "dbo.Suppliers";`
5.  Find the average unit price of products in the 'Beverages' category.
    `SELECT AVG(UnitPrice) FROM "dbo.Products" WHERE CategoryID = (SELECT CategoryID FROM "dbo.Categories" WHERE CategoryName = 'Beverages');`
6.  List the names and titles of all employees.
    `SELECT FirstName, LastName, Title FROM "dbo.Employees";`
7.  Find the total quantity of product with ProductID 11 ordered.
    `SELECT SUM(Quantity) FROM "dbo.Order Details" WHERE ProductID = 11;`
8.  List the CompanyName of shippers and their Phone numbers.
    `SELECT CompanyName, Phone FROM "dbo.Shippers";`
9.  Find the number of orders placed by customer 'VINET'.
    `SELECT COUNT(*) FROM "dbo.Orders" WHERE CustomerID = 'VINET';`
10. List the ProductName and UnitsInStock for all products where UnitsInStock is less than 10.
    `SELECT ProductName, UnitsInStock FROM "dbo.Products" WHERE UnitsInStock < 10;`
11. Find the total freight for all orders shipped to 'USA'.
    `SELECT SUM(Freight) FROM "dbo.Orders" WHERE ShipCountry = 'USA';`
12. List the ContactName of customers and their City, ordered by City.
    `SELECT ContactName, City FROM "dbo.Customers" ORDER BY City;`
13. Find the number of products supplied by supplier with SupplierID 1.
    `SELECT COUNT(*) FROM "dbo.Products" WHERE SupplierID = 1;`
14. List the OrderID and ShippedDate for orders that were shipped in 1997.
    `SELECT OrderID, ShippedDate FROM "dbo.Orders" WHERE ShippedDate LIKE '1997%';`
15. Find the total revenue from all orders (sum of UnitPrice * Quantity * (1 - Discount) from "dbo.Order Details").
    `SELECT SUM(UnitPrice * Quantity * (1 - Discount)) FROM "dbo.Order Details";`
16. List the ProductName and CategoryName for all products.
    `SELECT p.ProductName, c.CategoryName FROM "dbo.Products" p JOIN "dbo.Categories" c ON p.CategoryID = c.CategoryID;`
17. Find the number of orders handled by each employee, showing EmployeeID and the count.
    `SELECT EmployeeID, COUNT(*) FROM "dbo.Orders" GROUP BY EmployeeID;`
18. List the CompanyName of customers who have placed orders, and the OrderID for each order.
    `SELECT c.CompanyName, o.OrderID FROM "dbo.Customers" c JOIN "dbo.Orders" o ON c.CustomerID = o.CustomerID;`
19. Find the total value of orders for each customer, showing CustomerID and the sum of (UnitPrice * Quantity).
    `SELECT o.CustomerID, SUM(od.UnitPrice * od.Quantity) AS TotalOrderValue FROM "dbo.Orders" o JOIN "dbo.Order Details" od ON o.OrderID = od.OrderID GROUP BY o.CustomerID;`
20. List the ProductName, Supplier's CompanyName, and CategoryName for all products.
    `SELECT p.ProductName, s.CompanyName AS SupplierCompanyName, c.CategoryName FROM "dbo.Products" p JOIN "dbo.Suppliers" s ON p.SupplierID = s.SupplierID JOIN "dbo.Categories" c ON p.CategoryID = c.CategoryID;`

## Run the agent

### Subtask:
Run the agent with the updated tool and query to get the answer based on your dataset.


**Reasoning**:
The previous step successfully identified the correct query and executed it. Now, I will run the agent with the corrected query to get the final answer.



In [48]:
# new_query = "SELECT c.CompanyName, SUM(od.UnitPrice * od.Quantity * (1 - od.Discount)) AS TotalRevenue FROM dbo.Customers c JOIN dbo.\"Orders\" o ON c.CustomerID = o.CustomerID JOIN dbo.\"Order Details\" od ON o.OrderID = od.OrderID GROUP BY c.CompanyName ORDER BY TotalRevenue DESC LIMIT 10;"
from google.colab import userdata
import os

# Try to get the token from Colab secrets
hf_token = userdata.get('HF_TOKEN')

# If the token is not found in secrets, try from environment variables (less secure)
if hf_token is None:
    hf_token = os.environ.get('HF_TOKEN')

agent = CodeAgent(
    tools=[sql_engine],
    model=InferenceClientModel(model_id="meta-llama/Llama-2-7b-chat-hf", token=hf_token),
)

agent.run("Find the top 10 customers by total revenue")

╭──────────────────────────────────────────────────── New run ────────────────────────────────────────────────────╮
│                                                                                                                 │
│ Find the top 10 customers by total revenue                                                                      │
│                                                                                                                 │
╰─ InferenceClientModel - meta-llama/Llama-2-7b-chat-hf ──────────────────────────────────────────────────────────╯

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Step 1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Error in generating model output:
401 Client Error: Unauthorized for url: 
https://huggingface.co/api/models/meta-llama/Llama-2-7b-chat-hf?expand=inferenceProviderMapping (Request ID: 
Root=1-68f8e087-5c1f10cb6faed7ae4917ef26;fd2ea9aa-0315-4163-ae68-a54c96346032)

Invalid credentials in Authorization header

[Step 1: Duration 0.06 seconds]

AgentGenerationError: Error in generating model output:
401 Client Error: Unauthorized for url: https://huggingface.co/api/models/meta-llama/Llama-2-7b-chat-hf?expand=inferenceProviderMapping (Request ID: Root=1-68f8e087-5c1f10cb6faed7ae4917ef26;fd2ea9aa-0315-4163-ae68-a54c96346032)

Invalid credentials in Authorization header

## Summary:

### Q&A
What are the top 10 customers by total revenue?
The top 10 customers by total revenue are: 'Suprêmes délices' (\$3597.9), 'Frankenversand' (\$3536.6), 'Ernst Handel' (\$3488.68), 'Hanari Carnes' (\$2997.4), 'QUICK-Stop' (\$2555.68), 'Richter Supermarkt' (\$2490.5), 'Berglunds snabbköp' (\$2102.0), 'Rattlesnake Canyon Grocery' (\$2040.0), 'Toms Spezialitäten' (\$1863.4), and 'Wartian Herkku' (\$1722.56).

### Data Analysis Key Findings
* The dataset was provided as a zip file containing multiple .csv and .columns files.
* The .csv files were successfully extracted and loaded into pandas DataFrames.
* The database schema was dynamically created based on the columns and data types of the loaded DataFrames.
* Data from the DataFrames was successfully inserted into the newly created database tables.
* The SQL query to find the top 10 customers by total revenue required quoting table names with "dbo." prefixes to be compatible with the SQLite engine.

### Insights or Next Steps
* The process successfully demonstrated how to load data from a zip file, create a database schema dynamically, and query the data using an agent.
* Future work could involve exploring other relationships and insights within the dataset, such as analyzing product performance or employee sales figures.
